ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/API.hs

```diff
@@ -0,0 +1,572 @@
+data SnapCounters = SnapCounters {
+    -- | When was the last time we made a snapshot
+    prevSnapshotTime      :: !(Maybe Time)
+    -- | How many blocks have we processed since the last snapshot
+  , ntBlocksSinceLastSnap :: !Word64
+  }
+
+{-------------------------------------------------------------------------------
+  Validation
+-------------------------------------------------------------------------------}
+
+-- | When validating a sequence of blocks, these are the possible outcomes.
+data ValidateResult m l blk =
+    ValidateSuccessful       (Forker m l blk)
+  | ValidateLedgerError      (AnnLedgerError m l blk)
+  | ValidateExceededRollBack ExceededRollback
+
+type ValidateResult' m blk = ValidateResult m (ExtLedgerState blk) blk
+
+{-------------------------------------------------------------------------------
+  An annotated ledger error
+-------------------------------------------------------------------------------}
+
+-- | Annotated ledger errors
+data AnnLedgerError m l blk = AnnLedgerError {
+        -- | The ledger DB just /before/ this block was applied
+      annLedgerState  :: Forker m l blk
+
+      -- | Reference to the block that had the error
+    , annLedgerErrRef :: RealPoint blk
+
+      -- | The ledger error itself
+    , annLedgerErr    :: LedgerErr l
+    }
+
+type AnnLedgerError' m blk = AnnLedgerError m (ExtLedgerState blk) blk
+
+{-------------------------------------------------------------------------------
+  Trace validation events
+-------------------------------------------------------------------------------}
+
+newtype PushStart blk = PushStart { unPushStart :: RealPoint blk }
+  deriving (Show, Eq)
+
+newtype PushGoal blk = PushGoal { unPushGoal :: RealPoint blk }
+  deriving (Show, Eq)
+
+newtype Pushing blk = Pushing { unPushing :: RealPoint blk }
+  deriving (Show, Eq)
+
+data TraceValidateEvent blk =
+    -- | Event fired when we are about to push a block to a forker
+      StartedPushingBlockToTheLedgerDb
+        !(PushStart blk)
+        -- ^ Point from which we started pushing new blocks
+        (PushGoal blk)
+        -- ^ Point to which we are updating the ledger, the last event
+        -- StartedPushingBlockToTheLedgerDb will have Pushing and PushGoal
+        -- wrapping over the same RealPoint
+        !(Pushing blk)
+        -- ^ Point which block we are about to push
+  deriving (Show, Eq, Generic)
+
+{-------------------------------------------------------------------------------
+  Forker events
+-------------------------------------------------------------------------------}
+
+data TraceForkerEventWithKey =
+  TraceForkerEventWithKey ForkerKey TraceForkerEvent
+  deriving (Show, Eq)
+
+data TraceForkerEvent =
+    ForkerOpen
+  | ForkerCloseUncommitted
+  | ForkerCloseCommitted
+  | ForkerReadTablesStart
+  | ForkerReadTablesEnd
+  | ForkerRangeReadTablesStart
+  | ForkerRangeReadTablesEnd
+  | ForkerReadStatistics
+  | ForkerPushStart
+  | ForkerPushEnd
```

Comment:
It would output such a huge amount of data that the log file would grow beyond expectations. It might not exactly crash the node but it will have a huge impact on the performance.